mod/quiz/accessrule/seb/classes/external/validate_keys.php

```diff
@@ -0,0 +1,132 @@
+           'configkey' => new external_value(PARAM_ALPHANUMEXT, 'SEB config key',
+                VALUE_DEFAULT, null),
+           'browserexamkey' => new external_value(PARAM_ALPHANUMEXT, 'SEB browser exam key',
+                VALUE_DEFAULT, null),
+        ]);
+    }
+
+    /**
+     * Validate a SEB config key or browser exam key.
+     *
+     * @param string $cmid Course module ID.
+     * @param string $url URL of the page on which the SEB JS API generated the keys.
+     * @param string|null $configkey A SEB config key hash. Includes URL in the hash.
+     * @param string|null $browserexamkey A SEB browser exam key hash. Includes the URL in the hash.
+     * @return array
+     */
+    public static function execute(string $cmid, string $url, ?string $configkey = null, ?string $browserexamkey = null): array {
+        list(
+                'cmid' => $cmid,
+                'url' => $url,
+                'configkey' => $configkey,
+                'browserexamkey' => $browserexamkey
+            ) = self::validate_parameters(self::execute_parameters(), [
+            'cmid' => $cmid,
+            'url' => $url,
+            'configkey' => $configkey,
+            'browserexamkey' => $browserexamkey,
+        ]);
+
+        self::validate_context(\context_module::instance($cmid));
+
+        // At least one SEB key must be provided.
+        if (empty($configkey) && empty($browserexamkey)) {
+            throw new invalid_parameter_exception(get_string('error:ws:nokeyprovided', 'quizaccess_seb'));
+        }
+
+        // Check quiz exists corresponding to cmid.
+        if (($quizid = self::get_quiz_id($cmid)) === 0) {
+            throw new invalid_parameter_exception(get_string('error:ws:quiznotexists', 'quizaccess_seb', $cmid));
+        }
+
+        $accessmanager = new \quizaccess_seb\access_manager(quiz::create($quizid));
+
+        return ['valid' => $accessmanager->validate_seb_keys($configkey, $browserexamkey, $url)];
+    }
+
+    /**
+     * External function returns.
+     *
+     * @return external_single_structure
+     */
+    public static function execute_returns(): external_single_structure {
+        return new external_single_structure([
+            'valid' => new external_value(PARAM_BOOL, 'Is a provided key valid?',
+                    VALUE_REQUIRED, 0, NULL_NOT_ALLOWED)
+        ]);
+    }
+
+    /**
+     * Check if there is a valid quiz corresponding to a course module it.
+     *
+     * @param string $cmid Course module ID.
+     * @return int Returns quiz id if cmid matches valid quiz, or 0 if there is no match.
+     */
+    private static function get_quiz_id(string $cmid): int {
+        global $DB;
+        // Returns quizid if cmid matches a valid quiz.
+        $sql = "
+            SELECT cm.instance
+              FROM {course_modules} cm
+              JOIN {modules} m on m.id = cm.module
+             WHERE cm.id = :cmid
+               AND m.name = 'quiz'
```

Comment:
this may need to be moved to $params to be 100% sure that it would work in all DBs. Pretty sure it will, but flagging it here just in case.

Reply:
Moving it to param. Looks like what Moodle does most of the time